From 0883aca5ca59bd5bf496551909d93e25a54bc856 Mon Sep 17 00:00:00 2001
From: Katsuya Matsubara <matsu@igel.co.jp>
Date: Thu, 8 Nov 2012 16:37:03 +0900
Subject: [PATCH 75/97] ext/dfbvideosink: properly draw an image to non
 16-byte aligned address through MERAM

MERAM must be aligned on 16-byte boundaries. When dealing with non-16 byte
aligned address, start the MERAM write from the previous 16-byte boundary and
offset the VIO start address to account for the change.
---
 ext/directfb/dfbvideosink.c |   21 ++++++++++++++-------
 1 file changed, 14 insertions(+), 7 deletions(-)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index cd08eb7..b45f988 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -2081,6 +2081,8 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
 #if defined(HAVE_SHMERAM)
   phys[DST] = uiomux_all_virt_to_phys (dest_addr);
   if (phys[DST]) {
+    gulong py_frac = phys[DST] & 0x0000000f;
+
     viosurface[DST].pitch = 0;
     viosurface[DST].bpitchy = 4096;
     viosurface[DST].bpitchc = 4096;
@@ -2093,19 +2095,24 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
                     dst_format) - 1) & 0x7fff));
 
     viosurface[DST].py =
-        (void *) meram_get_icb_address (dfbvideosink->meram,
-        dfbvideosink->icby[DST], 0);
+        (void *) (meram_get_icb_address (dfbvideosink->meram,
+            dfbvideosink->icby[DST], 0) + py_frac);
 
     if (dfbvideosink->icbc[DST]) {
+      gulong physc, pc_frac;
       gulong chroma_byte_offset = 0;
+
+      physc = phys[DST] + chroma_byte_offset;
+      pc_frac = physc & 0x0000000f;
+
       meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST], MExxBSIZE,
           (((dst->h / 2 - 1) & 0x1fff) << 16) | ((pixel2byte (dst->w,
                       dst_format) - 1) & 0x7fff));
-      meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST],
-          MExxSSARA, phys[DST] + chroma_byte_offset);
-      viosurface[DST].pc = (void *)
-          meram_get_icb_address (dfbvideosink->meram, dfbvideosink->icbc[DST],
-          0);
+      meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST], MExxSSARA,
+          physc);
+      viosurface[DST].pc =
+          (void *) (meram_get_icb_address (dfbvideosink->meram,
+              dfbvideosink->icbc[DST], 0) + pc_frac);
     } else {
       viosurface[DST].pc = 0;
     }
-- 
1.7.10.4

